### DISCUSSION TITLE
[Unknown Mod][Game] Memorize the last used page of each hero's magic book

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#6928] [Ideas]
Always memorize the last used page of each hero's magic book, even for several combats (also on different days), and for the same day and several days on the adventure map for adventure spells page.
Currently, the last used page is only memorized for one combat, and not memorized at all on the adventure map.

This is not an option.
This can be implemented directly in fheroes2.

**Comment by Branikolog** [comment]
Hi, @LeHerosInconnu 
I personally OK with the implementation when the page is memorized only within a single battle. If I'm not mistaken this implementation was made intentionally 1-2 years ago, when this behaviour was introduced in fheroes2 engine.
I think each new battle has unique conditions so it looks natural for me, to open the book from the first page.
Memorizing a page mostly helps during a single situation when player, chooses a spell, cancels it to check something and then use it again. Or a player just misclicked and wants to enter spell book again to initiate the needed spell.
The same behaviour I expect for adventure map spells: IT would be nice to memorize page for a single day, but after a players ends turn next day he would browse his book from the first page.
I also believe such behaviour would help a player not to forget, which spells has a particular hero.

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

Let's say a player has dedicated one of his heroes to cast a Set Water Guardian spell on all the mines he owns.
In the hero's magic book, this spell is located on the third page.

In fheroes2:
![Memorize spell page 001](https://user-images.githubusercontent.com/43583994/232022430-47ccd88f-86a3-4c34-a1ea-802b4becafa5.png)
![Memorize spell page 002](https://user-images.githubusercontent.com/43583994/232022439-f2e7652b-65bf-4319-a740-840b2583dce2.png)

This means that on each new turn (each new day), the player will have to go from the first two pages of the magic book to the third page of the magic book.
However, if the last page used is memorized, this annoying action is not necessary.
And it can be even worse for combat spells.
With for example a hero dedicated to harvesting skeletons with a spell located in the fifth page of the magic book.

**Reply by Branikolog** [reply]
Hi, @LeHerosInconnu 
I could hardly imagine harvesting multiple times in a row or setting guardians. Especially within several days. :)
Several days ago I watched an experienced H2 streamer, who forgot, what spells he had in his book and was confused by the fact, spellbook was opened on the last page. He even asked, who stole his spells. :) So this happened within a single battle! What if the next battle he open a spellbook and watch just a few spells (which are simply last ones in his book)? He would definitely forget about the previous pages, since there's no visual representation, that the spellbook is opened on non-first page.

**Reply by LeHerosInconnu** [reply]
Hello @Branikolog,

> Hi, @LeHerosInconnu I could hardly imagine harvesting multiple times in a row or setting guardians. Especially within several days. :) Several days ago I watched an experienced H2 streamer, who forgot, what spells he had in his book and was confused by the fact, spellbook was opened on the last page. He even asked, who stole his spells. :) So this happened within a single battle! What if the next battle he open a spellbook and watch just a few spells (which are simply last ones in his book)? He would definitely forget about the previous pages, since there's no visual representation, that the spellbook is opened on non-first page.

Maybe the user could set this in a settings file.
Or during the game, left-clicking on the bookmark of combat spells (or adventure spells) on the desired page in the Magic Book switches on the memorization mode of the last used page, another left click switches off the memorization mode of the last used page.
It is visually indicated whether the page memorization mode is on or off.

**Edit 2023/05/06.**

The images are for explanation only.
To indicate the first and last page I had thought of darker corners.
In the example, there are only two pages, so the two corners are darker.
![Memorize spell page coins foncés 001](https://user-images.githubusercontent.com/43583994/236583720-72e46564-d655-41d9-8c64-d8eeac3bfd4a.png)

For the player's memorization of the page, I also thought to indicate this by a yellow border on the horned corner of the page.
![Memorize spell page coins foncés page mémorisée 001](https://user-images.githubusercontent.com/43583994/236583740-601db3ee-4c92-4d84-9dc2-99070c727c5a.png)

During the game, left-clicking directly on the horned corner (which at the same time indicates the on or off status of the memorisation) of the page of combat spells (or adventure spells) on the desired page in the Magic Book would be better than left-clicking on the bookmark.
